### PR TITLE
Adjust adjust unique hash table size to decrease memory bound .

### PR DESCRIPTION
**Pull request by zhaozheng09**
I found a small performance optimization point about unique_op.
>My test data: Input: size:20\*1000k, uniqued size:5\*1000k.

> test A(Origin):
>> Origin: The initial size of the hash table in unique_op is 2*input_size
>> Origin: spend: 2.517s

> I find heavy DRAM Bound by vtune .
> so I decrease hash size from default value(2*input_size) to custom value(1*input).

> test B(My test):
>> My test: The initial size of the hash table in unique_op is 1*input_size(change source code"tensorflow/core/kernels/unique_op.cc").
>> My test: spend: 1.709s**(↑32.1%)**

So I think we should adjust default hash table size from default value to dynamic size.

### CODE REVIEW

**Comment by zhaozheng09** [comment]
https://github.com/tensorflow/tensorflow/issues/47947

**Comment by zhaozheng09** [comment]
Data size and data repetition rate should be stable in every steps.  My codes should be used in stable environment .
My change maybe get bad performance in unstable data repetition. so I add a switch(default: use 2 * input_size), that will ensure stability(switch:2*input_size) and improve performance(Heuristic adjust hash size).

**Comment by zhaozheng09** [comment]
link issue:https://github.com/tensorflow/tensorflow/issues/47947

File: tensorflow/core/kernels/unique_op.cc [hunk 1]
```diff
@@ -76,7 +76,13 @@ struct UniqueOpHashMap<bfloat16, TIndex> {
 template <typename T, typename TIndex>
 class UniqueOp : public OpKernel {
  public:
-  explicit UniqueOp(OpKernelConstruction* context) : OpKernel(context) {}
+  static constexpr size_t INIT_SIZE = 1 << 12;
```
Comment (jpienaar) [inline tensorflow/core/kernels/unique_op.cc:79]: This should follow Google naming conventions (https://google.github.io/styleguide/cppguide.html)

File: tensorflow/core/kernels/unique_op.cc [hunk 2]
```diff
@@ -76,7 +76,13 @@ struct UniqueOpHashMap<bfloat16, TIndex> {
...
+
+ public:
```
Comment (jpienaar) [inline tensorflow/core/kernels/unique_op.cc:81]: This is already in a public section and does not need to be repeated

File: tensorflow/core/ops/array_ops.cc [hunk 3]
```diff
@@ -1491,6 +1491,7 @@ REGISTER_OP("Unique")
     .Output("idx: out_idx")
     .Attr("T: type")
     .Attr("out_idx: {int32, int64} = DT_INT32")
+    .Attr("auto_adjust_uniq_table_size: bool = false")
```
Comment (jpienaar) [inline tensorflow/core/ops/array_ops.cc:1494]: I'm not pro exposing what should be an implementation detail as an attribute. I'd rather this be done under the covers without the user having to choose. If the user needs to choose, this could be a separate op too. But that would require some more exhaustive benchmarking.

File: tensorflow/core/kernels/unique_op.cc [hunk 4]
```diff
@@ -237,6 +248,19 @@ class UniqueOp : public OpKernel {
...
+  void AutoAdjustUniqTableSize(const size_t uniq_size) {
+    if ((uniq_size << 1) > uniq_table_size_) {
+      uniq_table_size_ = uniq_size * 2.1;
+    } else if (uniq_size > (uniq_table_size_ << 2)) {
+      uniq_table_size_ = (uniq_size >> 2);
+    }
+  }
+
+ private:
+  size_t uniq_table_size_ = 0;
```
Comment (mrry) [inline tensorflow/core/kernels/unique_op.cc:262]: Drive-by comment: Multiple threads could read and write this field during `Compute()` if there are concurrent `Run()` calls, so you'll at least need to guard it with a mutex (and I suspect that it's a hot enough path that you might want to put the effort into implementing a lock-free version using `std::atomic`).

**Comment by gbaned** [comment]
@zhaozheng09 Can you please check @mrry's comments and keep us posted ? Thanks!

**Comment by zhaozheng09** [comment]
> @zhaozheng09 Can you please check @mrry's, @jpienaar's comments and keep us posted ? Thanks!

Ok, I've been a little busy at work these two days, I will modify it in the near future. thx~

File: tensorflow/core/kernels/unique_op.cc (hunk 4 above)
Comment (zhaozheng09) [reply tensorflow/core/kernels/unique_op.cc:262]: I copy code from tensorflow/core/kernels/dynamic_partition_op.cc
DynamicPartitionOp::num_partitions_ is a free-lock member variable .

**Comment by gbaned** [comment]
@zhaozheng09  Any update on this PR? Please. Thanks!
